Submitted by:            OpenMandriva
Date:                    2026-07-04
Initial Package Version: 1.4.4
Upstream Status:         Pending
Origin:                  Local fix for poppler 26.07.0 API removals
Description:             Fixes building Inkscape 1.4.4 with poppler 26.07 and
                         later. Poppler removed Object::streamGetDict() and
                         Object::arrayGetNF(), and changed Catalog::indexToLabel
                         to take std::string* instead of GooString* (GooString
                         now privately inherits std::string).

--- a/src/extension/internal/pdfinput/poppler-transition-api.h
+++ b/src/extension/internal/pdfinput/poppler-transition-api.h
@@ -15,6 +15,14 @@
 #include <glib/poppler-features.h>
 #include <poppler/UTF.h>
 
+#if POPPLER_CHECK_VERSION(26, 7, 0)
+#define _POPPLER_STREAM_GET_DICT(obj) (obj).getStream()->getDict()
+#define _POPPLER_ARRAY_GET_NF(obj, i) (obj).getArray()->getNF(i)
+#else
+#define _POPPLER_STREAM_GET_DICT(obj) (obj).streamGetDict()
+#define _POPPLER_ARRAY_GET_NF(obj, i) (obj).arrayGetNF(i)
+#endif
+
 #if POPPLER_CHECK_VERSION(26, 6, 0)
 #define _POPPLER_GET_GRAY(color, gray) getGray(color, gray)
 #define _POPPLER_GET_RGB(color, rgb) getRGB(color, rgb)
--- a/src/extension/internal/pdfinput/pdf-parser.cpp
+++ b/src/extension/internal/pdfinput/pdf-parser.cpp
@@ -293,9 +293,15 @@
     if (page) {
         // Increment the page building here and set page label
         Catalog *catalog = pdf_doc->getCatalog();
+#if POPPLER_CHECK_VERSION(26, 7, 0)
+        std::string label;
+        catalog->indexToLabel(page->getNum() - 1, &label);
+        builder->pushPage(getString(label), state);
+#else
         GooString *label = new GooString("");
         catalog->indexToLabel(page->getNum() - 1, label);
         builder->pushPage(getString(label), state);
+#endif
     }
 
     // Must come after pushPage!
@@ -818,7 +824,7 @@
             }
             _POPPLER_FREE(obj3);
             if (_POPPLER_CALL_ARGS_DEREF(obj3, obj2.dictLookup, "G").isStream()) {
-                if (_POPPLER_CALL_ARGS_DEREF(obj4, obj3.streamGetDict()->lookup, "Group").isDict()) {
+                if (_POPPLER_CALL_ARGS_DEREF(obj4, _POPPLER_STREAM_GET_DICT(obj3)->lookup, "Group").isDict()) {
                     std::unique_ptr<GfxColorSpace> blendingColorSpace;
                     GBool isolated = gFalse;
                     GBool knockout = gFalse;
@@ -881,7 +887,7 @@
   }
 
   // get stream dict
-  dict = str->streamGetDict();
+  dict = _POPPLER_STREAM_GET_DICT(*str);
 
   // check form type
   _POPPLER_CALL_ARGS(obj1, dict->lookup, "FormType");
@@ -2378,7 +2384,7 @@
     }
 
 //add layer at root if xObject has type OCG
-    _POPPLER_CALL_ARGS(obj2, obj1.streamGetDict()->lookup, "OC");
+    _POPPLER_CALL_ARGS(obj2, _POPPLER_STREAM_GET_DICT(obj1)->lookup, "OC");
     if(obj2.isDict()){
         auto type_dict = obj2.getDict();
         if (type_dict->lookup("Type").isName("OCG")) {
@@ -2396,7 +2402,7 @@
         }
     }
 
-    _POPPLER_CALL_ARGS(obj2, obj1.streamGetDict()->lookup, "Subtype");
+    _POPPLER_CALL_ARGS(obj2, _POPPLER_STREAM_GET_DICT(obj1)->lookup, "Subtype");
     if (obj2.isName(const_cast<char*>("Image"))) {
         _POPPLER_CALL_ARGS(refObj, res->lookupXObjectNF, name);
         doImage(&refObj, obj1.getStream(), gFalse);
@@ -2404,7 +2410,7 @@
     } else if (obj2.isName(const_cast<char*>("Form"))) {
         doForm(&obj1);
     } else if (obj2.isName(const_cast<char*>("PS"))) {
-        _POPPLER_CALL_ARGS(obj3, obj1.streamGetDict()->lookup, "Level1");
+        _POPPLER_CALL_ARGS(obj3, _POPPLER_STREAM_GET_DICT(obj1)->lookup, "Level1");
     } else if (obj2.isName()) {
         error(errSyntaxError, getPos(), "Unknown XObject subtype '{0:s}'", obj2.getName());
     } else {
@@ -2597,7 +2603,7 @@
 	            goto err1;
             }
             maskStr = smaskObj.getStream();
-            maskDict = smaskObj.streamGetDict();
+            maskDict = _POPPLER_STREAM_GET_DICT(smaskObj);
             _POPPLER_CALL_ARGS(obj1, maskDict->lookup, "Width");
             if (obj1.isNull()) {
                     _POPPLER_FREE(obj1);
@@ -2675,7 +2681,7 @@
                 goto err1;
             }
             maskStr = maskObj.getStream();
-            maskDict = maskObj.streamGetDict();
+            maskDict = _POPPLER_STREAM_GET_DICT(maskObj);
             _POPPLER_CALL_ARGS(obj1, maskDict->lookup, "Width");
             if (obj1.isNull()) {
                 _POPPLER_FREE(obj1);
@@ -2775,7 +2781,7 @@
     }
 
     // get stream dict
-    dict = str->streamGetDict();
+    dict = _POPPLER_STREAM_GET_DICT(*str);
 
     // check form type
     _POPPLER_CALL_ARGS(obj1, dict->lookup, "FormType");
--- a/src/extension/internal/pdfinput/poppler-utils.cpp
+++ b/src/extension/internal/pdfinput/poppler-utils.cpp
@@ -191,7 +191,7 @@
             n = obj->arrayGetLength();
             h->hash((char *)&n, sizeof(int));
             for (i = 0; i < n; ++i) {
-                const Object &obj2 = obj->arrayGetNF(i);
+                const Object &obj2 = _POPPLER_ARRAY_GET_NF(*obj, i);
                 hashFontObject1(&obj2, h);
             }
             break;
@@ -545,7 +545,7 @@
                 continue;
 
             Ref resourcesRef;
-            const Object resObj = obj2.streamGetDict()->lookup("Resources", &resourcesRef);
+            const Object resObj = _POPPLER_STREAM_GET_DICT(obj2)->lookup("Resources", &resourcesRef);
             if (resourcesRef != Ref::INVALID() && !visitedObjects.insert(resourcesRef.num).second)
                 continue;
 
